executor/new_join.go

```diff
@@ -0,0 +1,253 @@
+)
+
+type orderedRow struct {
+	key []byte
+	row *Row
+}
+
+type orderedRows []orderedRow
+
+// Len returns the number of rows.
+func (e orderedRows) Len() int {
+	return len(e)
+}
+
+// Swap implements sort.Interface Swap interface.
+func (e orderedRows) Swap(i, j int) {
+	e[i], e[j] = e[j], e[i]
+}
+
+// Less implements sort.Interface Less interface.
+func (e orderedRows) Less(i, j int) bool {
+	return bytes.Compare(e[i].key, e[j].key) < 0
+}
+
+// IndexLookUpJoin fetches batches of data from outer executor and constructs ranges for inner executor.
+type IndexLookUpJoin struct {
+	baseExecutor
+
+	innerExec DataReader
+
+	cursor      int
+	resultRows  []*Row
+	outerRows   orderedRows
+	innerRows   orderedRows
+	innerDatums orderedRows // innerDatums are extracted by innerRows and innerJoinKeys
+	exhausted   bool        // exhausted means whether all data has been extracted
+
+	outerJoinKeys   []*expression.Column
+	innerJoinKeys   []*expression.Column
+	leftConditions  expression.CNFExprs
+	rightConditions expression.CNFExprs
+	otherConditions expression.CNFExprs
+	defaultValues   []types.Datum
+	outer           bool
+}
+
+// Open implements the Executor Open interface.
+func (e *IndexLookUpJoin) Open() error {
+	e.cursor = 0
+	e.exhausted = false
+	return errors.Trace(e.children[0].Open())
+}
+
+// Close implements the Executor Close interface.
+func (e *IndexLookUpJoin) Close() error {
+	e.resultRows = nil
+	e.outerRows = nil
+	e.innerDatums = nil
+	return errors.Trace(e.children[0].Close())
+}
+
+// Next implements the Executor Next interface.
+// We will fetch batches of row from outer executor, construct the inner datums and sort them.
+// At the same time we will fetch the inner row by the inner datums and apply merge join.
+func (e *IndexLookUpJoin) Next() (*Row, error) {
+	for e.cursor == len(e.resultRows) {
+		if e.exhausted {
+			return nil, nil
+		}
+		batchSize := e.ctx.GetSessionVars().IndexLookupSize
+		e.outerRows = e.outerRows[:0]
+		e.resultRows = e.resultRows[:0]
+		e.innerDatums = e.innerDatums[:0]
+		for i := 0; i < batchSize; i++ {
+			outerRow, err := e.children[0].Next()
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if outerRow == nil {
+				e.exhausted = true
+				break
+			}
+			match, err := expression.EvalBool(e.leftConditions, outerRow.Data, e.ctx)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if match {
+				joinDatums := make([]types.Datum, 0, len(e.outerJoinKeys))
+				for _, col := range e.outerJoinKeys {
+					datum, _ := col.Eval(outerRow.Data)
+					joinDatums = append(joinDatums, datum)
+				}
+				joinOuterEncodeKey, err := codec.EncodeValue(nil, joinDatums...)
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
+				e.outerRows = append(e.outerRows, orderedRow{key: joinOuterEncodeKey, row: outerRow})
+				e.innerDatums = append(e.innerDatums, orderedRow{key: joinOuterEncodeKey, row: &Row{Data: joinDatums}})
+			} else if e.outer {
```

Comment:
If `match` is false, why should we still remain the outerRow when it is an OuterJoin.